torch/_dynamo/variables/misc.py

```diff
@@ -224,8 +224,16 @@ def visit(node):
         VariableTracker.apply(visit, (args, kwargs))
 
         if requires_grad and torch.is_grad_enabled():
-            # TODO(jansel): handle this in training mode
-            unimplemented("autograd.Function with requires_grad")
+            from .torch import TorchHigherOrderOperator
+
+            def trampoline_autograd_fn(*args, **kwargs):
+                return self.fn_cls.apply(*args, **kwargs)
+
+            # Speculate fwd
+            # TODO(voz): Check bwd soundness, or something, I dunno, bug Horace
```

Comment:
We should probably check bwd soundness, or hide this PR behind a config. Otherwise people are going to come to us with silent correctness errors when using autograd.Function with dynamo

Reply:
bwd soundness is another can of worms though. As @bdhirsh mentions often, we don't really know if the grad_output pass to the backward are Tensors (or None), we don't know their sizes or strides, we don't know if they're Tensor subclasses, etc.

Reply:
Yes, I am not sure what to do about this - open to suggestions on how to do it. 

I suppose we could do something like speculate_graph on bwd, but throw out the results?

Reply:
Yeah we could assume  we get all the inputs, and if that's not the case fallback to backward fn?

Reply:
Here's a lightweight proposal from the comments above:
- follow what AOTAutograd does for grad_outputs. Last time I checked this was to assume the grad_outputs have the same features (size/stride/etc) as the outputs, and it adds coercion of the grad_outputs to the assumed size/stride/etc, but this might be out of date
- speculate_subgraph on bwd on grad_outputs created to be the same size/stride/etc, throw out the results after we've proven safety.